clippy_lints/src/invalid_build_cfg.rs

```diff
@@ -0,0 +1,308 @@
+use rustc_ast::ast::{Attribute, MacCall};
+use rustc_ast::token::{LitKind, TokenKind};
+use rustc_ast::tokenstream::{TokenStream, TokenTree};
+use rustc_ast::MetaItem;
+use rustc_errors::Applicability;
+use rustc_lint::{EarlyContext, EarlyLintPass, LintContext};
+use rustc_session::declare_lint_pass;
+use rustc_span::{sym, Span, Symbol};
+
+use clippy_utils::diagnostics::{span_lint, span_lint_and_sugg};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for usage of `#[cfg]`/`#[cfg_attr]`/`cfg!()` in `build.rs` scripts.
+    ///
+    /// ### Why is this bad?
+    /// It checks the `cfg` values for the *host*, not the target. For example, `cfg!(windows)` is
+    /// true when compiling on Windows, so it will give the wrong answer if you are cross compiling.
+    /// This is because build scripts run on the machine performing compilation, rather than on the
+    /// target.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// if cfg!(windows) {}
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// if std::env::var("CARGO_CFG_WINDOWS").is_ok() {}
+    /// ```
+    #[clippy::version = "1.80.0"]
+    pub INVALID_BUILD_CFG,
+    suspicious,
+    "invalid use of cfg in `build.rs`"
+}
+
+declare_lint_pass!(InvalidBuildCfg => [INVALID_BUILD_CFG]);
+
+#[derive(Debug)]
+enum CfgAst {
+    Os(Symbol),
+    Any(Vec<CfgAst>),
+    All(Vec<CfgAst>),
+    Not(Box<CfgAst>),
+    TargetKeyValue(Symbol, Symbol),
+    Feature(Symbol),
+    OtherTarget(Symbol, Symbol),
+}
```

Comment:
Some documentation would be nice, it's not immediately clear why `Feature`, `Os` and `OtherTarget` exists.